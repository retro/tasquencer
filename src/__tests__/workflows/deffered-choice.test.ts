import { Effect } from 'effect';
import { expect, it } from 'vitest';

import { Builder, IdGenerator, Service } from '../../index.js';
import { getEnabledTaskNames, makeIdGenerator } from '../shared.js';

it('supports the deferred choice pattern', () => {
  const workflowDefinition = Builder.workflow()
    .withName('deferred-choice')
    .startCondition('start')
    .task('task_1', Builder.emptyTask())
    .task('task_1a', Builder.emptyTask())
    .task('task_2', Builder.emptyTask())
    .task('task_2a', Builder.emptyTask())
    .endCondition('end')
    .connectCondition('start', (to) => to.task('task_1').task('task_2'))
    .connectTask('task_1', (to) => to.task('task_1a'))
    .connectTask('task_2', (to) => to.task('task_2a'))
    .connectTask('task_1a', (to) => to.condition('end'))
    .connectTask('task_2a', (to) => to.condition('end'));

  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    expect(yield* $(service.getState())).toMatchSnapshot();

    yield* $(service.start());
    const state1 = yield* $(service.getState());
    expect(state1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1)).toEqual(new Set(['task_1', 'task_2']));

    yield* $(service.startTask('task_1'));
    const state2 = yield* $(service.getState());
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['task_1a']));

    yield* $(service.startTask('task_1a'));
    const state3 = yield* $(service.getState());
    expect(state3).toMatchSnapshot();
    expect(state3.workflows[0]?.state).toBe('completed');
  });

  Effect.runSync(program);
});
