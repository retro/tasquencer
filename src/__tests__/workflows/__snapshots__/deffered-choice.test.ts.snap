// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`supports the deferred choice pattern 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_1->task_1a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_2->task_2a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "task_1",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_1a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "deferred-choice",
      "parent": null,
      "state": "initialized",
    },
  ],
}
`;

exports[`supports the deferred choice pattern 2`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_1->task_1a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_2->task_2a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "task_1",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_1a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "deferred-choice",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the deferred choice pattern 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 1,
      "name": "implicit:task_1->task_1a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_2->task_2a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "task_1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_1a",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "deferred-choice",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`supports the deferred choice pattern 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_1->task_1a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "marking": 0,
      "name": "implicit:task_2->task_2a",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "task_1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 1,
      "name": "task_1a",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
    {
      "generation": 0,
      "name": "task_2a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "deferred-choice",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "deferred-choice",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;
